### DISCUSSION TITLE
What should be featured in this template?

### DISCUSSION THREAD

**Discussion by malliaridis** [arkivanov/decompose-multiplatform-template#1] [Ideas]
I have started working on a proposal and was wondering which libraries / elements should be featured in this template.

As we already discussed in https://github.com/arkivanov/Decompose/issues/434, including a module for shared Decompose code and another module for shared Compose UI is one of the requested features. This includes two `ios` modules, one with Compose UI and one with SwiftUI. The same integration can be done for Android, providing one module with Compose UI and one with XML. Would that be worth the extra module?

And what about other integrations, like [MVIKotlin](https://github.com/arkivanov/MVIKotlin/), state preservation ([parcelize-darwin](com.arkivanov.parcelize.darwin)), Reaktive or Coroutines support?

In my opinion, a basic navigation example with state preservation would be sufficient, but I usually use Decompose together MVIKotlin and Coroutines and therefore I would find it handy to also have an extended template with those included.

**Comment by arkivanov** [comment]
Thanks for working on this!

I suggest to begin with the following module structure:

- `shared` - contains Decompose components, I think we can begin with just one `RootComponent` showing "Hello from Decompose".
- `compose-ui` - depends on `shared` and contains Multiplatform Compose UI, just one `RootContent` function to begin with.
- `app-android` - depends on both KMP modules and contains only Compose integration (I think no need for XML for now).
- `app-ios-compose` - an iOS app with Compose integration, depends on both KMP modules.
- `app-ios-swift` - an iOS app with SwiftUI integration, depends only on `shared` module.

What do you think?

**Reply by malliaridis** [reply]
Sounds simple and sufficient for the first template proposal. :) I will go for this then, but any additional ideas and opinions are welcome of course. :)

**Reply by malliaridis** [reply]
After experimenting and trying to figure out how to integrate two gradle modules into a single ios app, I finally managed to correctly load and run everything (sadly, `shared` and `compose-ui` cannot be used as separate pods and combined in Swift).

Before I proceed with the swift ap with only the `shared` module I wanted to get some input about the approach I chose.

Since the `cocoapods` gradle plugin uses a Podfile as parameter, I was thinking of creating a single `app-ios` module and add there two iOS modules, one for compose and one for shared only. Then, provide two targets in the Podfile like below:
```ruby
target 'app-ios-compose' do
  use_frameworks!
  platform :ios, '14.1'
  pod 'compose_ui', :path => '../compose-ui'
  # Pod compose_ui already contains and exports shared module
end

target 'app-ios-swift' do
  use_frameworks!
  platform :ios, '14.1'
  pod 'shared', :path => '../shared'
end
```

**Reply by arkivanov** [reply]
Please let's avoid `cocoapods` plugin. It's much simpler to use just the regular integration as it's done in Decompose samples. See the article: https://blog.jetbrains.com/kotlin/2021/07/multiplatform-gradle-plugin-improved-for-connecting-kmm-modules/.

I thought `app-ios-compose` could be a normal Gradle module? So you just add both modules as `implementation` dependencies. Otherwise, if we need an Xcode project for this, then we would need to an umbrella Gradle module for this.

**Reply by malliaridis** [reply]
I was planning to follow the format and structure of the Compose Multiplatform Template (more or less) which uses the `cocoapods` plugin.

One of the reasons I decided to go with that approach is that there are a couple of setup steps, hints and warnings about the setup for iOS and I wanted to reference those by linking to the original template. Another reason why I considered that template as the base is because many people may come from that template and want to adapt / integrate Decompose by looking up only the differences.

As an alternative I considered the sample project from Decompose, which I could merge it into one Multiplatform project and reduce the content to just a single `RootComponent` for providing the basic template project. This may be more straight-forward but could be less useful for people coming from Compose MPP directly.

Should I stick with the Decompose samples instead and avoid the `cocoapods` plugin?

**Reply by arkivanov** [reply]
Yes please, let's avoid Cocoapods for now. I truly believe that one should only use Cocoapds with KMP if it can't be avoided (e.g. you need to use a ObjC dependency in your shared Kotlin code). Also Cocapods is know to be less stable than the regular framework integration. E.g. multiple people had issues with exporting transitive dependencies to the shared framework, all were resolved by switching to regular framework.

**Reply by malliaridis** [reply]
I didn't expect it to be that easy without the `cocoapods` plugin and I didn't know that you could go with Kotlin only for iOS.

I definitely have to migrate my projects now. 😁 I will soon make a PR so you can review it and request changes. :)

**Reply by arkivanov** [reply]
Awesome! Please try making small PRs, so it's easier to review. Thanks for working on this! 🚀

**Reply by azlekov** [reply]
Hey guys, I'm following the discussions some and there because I'm also interested in unifying my stack, so I can bootstrap quickly some pet projects. I spend some time to adjust the jetbrains template to use decompose, no cocoapods, material3 - you can check it here: https://github.com/dodi-smart/decompose-multiplatform-ios-android-template

@malliaridis regarding migrating from cocoapods to framework, check this: https://github.com/JetBrains/compose-multiplatform/pull/3340

**Reply by azlekov** [reply]
@arkivanov can you review my changes there are in single commit. I'm concern about plugging the lifecycle and everything in iOS using the compose `UIViewControllerRepresentable`

**Reply by malliaridis** [reply]
@arkivanov I agree on keeping PRs small. It's just the initial one that will be a bit overwhelming. But any follow-up changes can be kept simple.

Thanks for the resources @azlekov. The PR you posted is quite similar to the blog post Arkadii shared. I realized that the easiest approach is to create a multiplatform project via Android Studio for iOS and Android and go from there (adding the other platforms one by one). That way bootstrapping a project without `cocoapods` is straight forward and works out of the box. I will also include many parts of the `sample` project the Decompose repo contains, since there are already a lot of best practices applied and it works best with Decompose.

But I will definitely look into your project too to see if I missed anything crucial. Feel free to review and give feedback to the proposal too. The more input and opinions the better the template can be optimized for the general purpose. :)

**Reply by arkivanov** [reply]
@azlekov Thanks for your input!

PS: I've left a comment in your repository.

**Comment by devjorgecastro** [comment]
> I have started working on a proposal and was wondering which libraries / elements should be featured in this template.
> 
> As we already discussed in [arkivanov/Decompose#434](https://github.com/arkivanov/Decompose/issues/434), including a module for shared Decompose code and another module for shared Compose UI is one of the requested features. This includes two `ios` modules, one with Compose UI and one with SwiftUI. The same integration can be done for Android, providing one module with Compose UI and one with XML. Would that be worth the extra module?
> 
> And what about other integrations, like [MVIKotlin](https://github.com/arkivanov/MVIKotlin/), state preservation ([parcelize-darwin](com.arkivanov.parcelize.darwin)), Reaktive or Coroutines support?
> 
> In my opinion, a basic navigation example with state preservation would be sufficient, but I usually use Decompose together MVIKotlin and Coroutines and therefore I would find it handy to also have an extended template with those included.


I suggest that we could have different examples; perhaps in another example, we could have the implementation with MVI. The idea is to have very simple and easy-to-understand resources to learn how to set up Decompose, especially for those of us who are starting with KMP (I say this from my perspective with the little time I have spent on this technology 🙂).

**Reply by arkivanov** [reply]
Integrating with other libraries may bloat the template and confuse users. E.g. one may think that MVIKotlin is necessary, while it's not. Though, later we could add MVIKotlin (and other libraries) in a separate branch, or maybe in a fork, or something like that.

**Reply by malliaridis** [reply]
I like the idea of providing various integrations in separate branches. And if the changes in the commits are kept clean, users will be able to lookup only the differences (e.g. with `git diff`), allowing them to integrate other libraries at any stage of the project with the help of this template.

**Comment by arkivanov** [comment]
I have added a CI setup, so that all future PRs are checked.

**Reply by malliaridis** [reply]
Nicely done. Just a quick question: Is the Java version 17 fine or should that match Java 11 like in the gradle files?

**Reply by arkivanov** [reply]
That should be fine. It shouldn't be lower than 11, though.
